src/python/pants/engine/target.py

```diff
@@ -1902,6 +1925,36 @@ def to_unparsed_address_inputs(self) -> UnparsedAddressInputs:
         return UnparsedAddressInputs(self.value or (), owning_address=self.address)
 
 
+class BinaryTargetDependencies(SpecialCasedDependencies):
+    """A base class for dependencies of a target that represent a binary artifact.
+
+    A "binary target" is one that:
+
+    A) Does not own any source code itself.
```

Comment:
Hm, but Docker targets do own their `Dockerfile` and it would be onerous to require `docker_library` like we do with `pex_binary`. This is tricky 🤔

Reply:
Currently dependency inference requires a regular `Dependencies` I think. So if we want to allow a package target to own sources we have to modify dep inference. 

Or we can make `docker_image` a macro that generates a `docker_library` and a `dockerfile` target, or something?